### PR TITLE
Allow stoa to call getPreimage and improve TestAgora

### PR DESCRIPTION
**Pull request by Geod24**
Pretty straightforward, see each commit for details.

### CODE REVIEW

File: src/modules/agora/AgoraClient.ts [hunk 1]
```diff
@@ -108,6 +108,32 @@ export class AgoraClient implements FullNodeAPI
                 });
         });
     }
+
+    /**
+     * Request the latest PreImageInfo for an enrollment key
+     *
+     * @param {enroll_key} Hash of the UTXO used by the validator
```
Comment (TrustHenry) [inline src/modules/agora/AgoraClient.ts]:
```suggestion
     * @param enroll_key{Hash} Hash of the UTXO used by the validator
```

Comment (TrustHenry) [reply src/modules/agora/AgoraClient.ts]: Please add @returns document.

Comment (Geod24) [reply src/modules/agora/AgoraClient.ts]: Looking at the official documentation, it should be `@param {type} name`

https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#param-and-returns

Comment (Geod24) [reply src/modules/agora/AgoraClient.ts]: Fixed documentation

File: tests/Utils.ts [hunk 2]
```diff
@@ -75,32 +75,52 @@ export class TestAgora
...
+            let max_blocks = Number(req.query.max_blocks);
+            if (max_blocks <= 0)
+            {
+                res.status(400).json({ errorMessage: 'Query parameter max_block must be strictly positive' });
+                return;
+            }
+
+            if (this.blocks.length < 1)
+            {
+                res.status(500).json({ errorMessage: 'No data on server' });
```
Comment (TrustHenry) [inline tests/Utils.ts]: `errorMessage:`
Now should we apply all response error messages key in common?

File: src/modules/agora/AgoraClient.ts [hunk 3]
```diff
@@ -108,6 +108,33 @@ export class AgoraClient implements FullNodeAPI
...
         });
     }
+
+    /**
+     * Request the latest PreImageInfo for an enrollment key
+     *
+     * @param enroll_key Hash of the UTXO used by the validator
+     * @returns A `PreImageInfo` matching this `enroll_key`
+     */
+    public getPreimage (enroll_key: Hash): Promise<PreImageInfo>
```
Comment (TrustHenry) [inline src/modules/agora/AgoraClient.ts:118]: Do you want to change the method of collecting pre-image data from `put` to `get` request?

**Review by TrustHenry — APPROVED** [review]
LGTM

File: src/modules/agora/AgoraClient.ts (hunk 3 above)
Comment (TrustHenry) [reply src/modules/agora/AgoraClient.ts:118]: Or is it just for testing?

File: src/modules/agora/AgoraClient.ts [hunk 4]
```diff
@@ -83,9 +83,9 @@ export class AgoraClient implements FullNodeAPI
      * @param block_height - The height of the block
      * @param max_blocks - The maximum number of block to request
      */
-    public getBlocksFrom (block_height: Height, max_blocks: number): Promise<Array<any>>
+    public getBlocksFrom (block_height: Height, max_blocks: number): Promise<Block[]>
```
Comment (MichaelKim20) [inline src/modules/agora/AgoraClient.ts]: Does it automatically convert to an instance of Block?

File: src/modules/agora/AgoraClient.ts (hunk 3 above)
Comment (Geod24) [reply src/modules/agora/AgoraClient.ts:118]: No, we keep Agora's push approach, this is here for a fallback.

File: tests/Utils.ts (hunk 2 above)
Comment (Geod24) [reply tests/Utils.ts]: No, I should have used `statusText`

Comment (MichaelKim20) [reply tests/Utils.ts]: A `statusMessage` is used in Agora.
https://github.com/bpfkorea/vibe.d/blob/master/web/vibe/web/rest.d#L1466

**Review by MichaelKim20 — COMMENTED** [review]
@Geod24 : That's a great suggestion. I requested some revisions. Please check.

**Comment by Geod24** [comment]
Comments fixed:
- The first commit was lacking and has been moved to #245 
- The second commit missed a `reviver` call, added it;
- Changed `errorMessage` to the correct `statusText`;

**Review by TrustHenry — APPROVED** [review]

**Comment by Geod24** [comment]
Anything else ?

**Review by MichaelKim20 — APPROVED** [review]
LGTM

**Pull request merged**